src/merkle/tiered_smt/context.rs

```diff
@@ -0,0 +1,203 @@
+use super::{
+    BTreeMap, BTreeSet, BitsIterator, ContentType, Felt, Leaf, NodeIndex, Rpo256, StarkField,
+    StorageError, TieredSmt, Vec, Word, ZERO,
+};
+
+pub struct TreeUpdateContext<'a> {
+    pub(super) tier_leaves: BTreeMap<NodeIndex, Leaf>,
+    pub(super) ordered_leaves: BTreeMap<NodeIndex, Vec<Leaf>>,
+    pub(super) removed_leaves: BTreeSet<NodeIndex>,
+    pub(super) nodes: BTreeMap<NodeIndex, Word>,
+    pub(super) tree: &'a TieredSmt,
+    pub(super) types: BTreeMap<NodeIndex, ContentType>,
+}
+
+impl<'a> From<&'a TieredSmt> for TreeUpdateContext<'a> {
+    fn from(tree: &'a TieredSmt) -> Self {
+        Self {
+            tier_leaves: BTreeMap::new(),
+            ordered_leaves: BTreeMap::new(),
+            removed_leaves: BTreeSet::new(),
+            nodes: BTreeMap::new(),
+            tree,
+            types: BTreeMap::new(),
+        }
+    }
+}
+
+impl<'a> TreeUpdateContext<'a> {
+    pub fn get_node(&self, index: &NodeIndex) -> Result<Word, StorageError> {
+        self.nodes
+            .get(index)
+            .map(|node| Ok(*node))
+            .unwrap_or_else(|| self.tree.get_node(index))
+    }
+
+    pub fn get_type(&self, index: &NodeIndex) -> Result<ContentType, StorageError> {
+        self.types
+            .get(index)
+            .map(|t| Ok(*t))
+            .unwrap_or_else(|| self.tree.get_type(index))
+    }
+
+    pub fn get_tier_leaf(&self, index: &NodeIndex) -> Result<Option<Leaf>, StorageError> {
+        if let Some(leaf) = self.tier_leaves.get(index).copied() {
+            return Ok(Some(leaf));
+        }
+        let leaf = self
+            .tree
+            .storage
+            .get_leaves(index)?
+            .unwrap_or_default()
+            .into_iter()
+            .next();
+        Ok(leaf)
+    }
+
+    pub fn get_ordered_leaves(&self, index: &NodeIndex) -> Result<Option<Vec<Leaf>>, StorageError> {
+        if let Some(leaves) = self.ordered_leaves.get(index).cloned() {
+            return Ok(Some(leaves));
+        }
+        match self.tree.storage.get_leaves(index)? {
+            Some(leaves) if leaves.is_empty() => Err(StorageError::LeafNodeWithoutDigest),
+            Some(leaves) => Ok(Some(leaves)),
+            None => Ok(None),
+        }
+    }
+
+    pub fn insert_leaf(&mut self, leaf: Leaf) -> Result<(), StorageError> {
+        // TODO this can be optimized to not recurse
+        let mut bits = BitsIterator::from(&leaf.key);
+        let mut index = BitsIterator::traverse_from_root(bits.by_ref());
+
+        // traverse until non-internal node
+        let mut current = self.get_type(&index)?;
+        while current.is_internal() {
+            index = match BitsIterator::traverse(index, bits.by_ref()) {
+                Some(idx) => idx,
+                None => return Err(StorageError::MaxDepthIsInternal),
+            };
+            current = self.get_type(&index)?;
+        }
```

Comment:
I don't think we need to do a full traversal here. Leaves can be only at depth 16, 32, 48, and 64 - so, we can just check these depths to figure out where the leaf needs to be inserted.

Reply:
I'm not sure that is possible. Let's assume we have a depth of `4` and a key `0b0101`

We start from the root, with index `{ depth:0, index: 0 }`
The final index should be `{ depth: 4, index: 10 }`

In order to get this `10`, I think the only way is to iterate the bits of the key and shift the index accordingly (left or right).

If it was the other way around (we have the index and we want to traverse one tier towards the root), then we just divide the index `8` and we are good.

Do we have some algorithm to perform this traverse without iterating the bits?

Reply:
Here is how I think about it. Let's say we have a 64-bit node index `a`. We can view this index as four 16-bit limbs: `[a0, a1, a2, a3]` where `a3` is the most significant limb. If we want to check if a node for index `a` is in a tree we need to do the following lookups:

* `{ depth = 16, index = a3 }`
* `{ depth = 32, index = a2 || a3 }`
* `{ depth = 48, index = a1 || a2 || a3 }`
* `{ depth = 64, index = a0 || a1 || a2 || a3 }` which is the same thing as `{ depth = 64, index = a }`

Here is a concrete example on a tree of depth 4 and with only 2 tiers (at depth 2 and 4). Such a tree would look something like this:

![image](https://user-images.githubusercontent.com/43513081/216460685-e418ee28-e17a-498d-97f2-99e0f7ed2cdf.png)

Let's say `a = 5` or `0101` in binary (big-endian order). The checks we need to do are:

* `{ depth = 2, index = 1 }` where 1 is the the numerical value of the two most significant bits of `a` (i.e., `01` in big-endian order). If the node there is an internal node, we need to check the next tier.
* `{ depth = 4, index = 5}` where 5 is the full index of node.

These two checks should be sufficient to figure out where and how to insert the new leaf, and we don't need to look at nodes at depths 1 and 3.

Reply:
Done! As discussed, the reason I used the LSB first was because I suspected we had non-uniform bits for the MSB of canonical representations of our field elements.

However, that is either not true or negligible, so this simplification is unquestionably an improvement.